monitor/plugins/spark_progress_upv.py

```diff
@@ -0,0 +1,163 @@
+
+        self.conn = paramiko.SSHClient()
+        self.conn.set_missing_host_key_policy(paramiko.AutoAddPolicy())
+        self.conn.connect(hostname=api.mesos_cluster_addr,
+                          username=api.mesos_username,
+                          password=api.mesos_password)
+
+        self.spark_id = self._discover_id_from_spark()
+
+    def _publish_measurement(self, jobs):
+
+        application_progress_error = {}
+
+        # Init
+        jobs.reverse()
+
+        if not len(jobs) == 0:
+            current_job = jobs[self.current_job_id]
+
+            if current_job['status'] == 'FAILED':
+                self.current_job_id = len(jobs) - 1
+
+            elif current_job['status'] == 'SUCCEEDED':
+                elapsed_time = float(self._get_elapsed_time(
+                               current_job['submissionTime']))
+
+                self.remaining_time = self.remaining_time - elapsed_time
+
+                self.current_job_id = len(jobs) - 1
+
+                # Job Time
+                self.job_expected_time = (self.remaining_time
+                                 / (float(self.number_of_jobs)
+                                 - float(self.current_job_id)))
+
+            elif current_job['status'] == 'RUNNING':
+                # Job Progress
+                job_progress = (current_job['numCompletedTasks']
+                                / float(current_job['numTasks']))
+
+                # Elapsed Time
+                elapsed_time = float(self._get_elapsed_time(
+                               current_job['submissionTime']))
+
+                # Reference Value
+                ref_value = (elapsed_time / self.job_expected_time)
+
+                # Error
+                error = job_progress - ref_value
+                
+                application_progress_error['name'] = ('application-progress'
+                                                      '.error')
+
+                application_progress_error['value'] = error
+                application_progress_error['timestamp'] = time.time() * 1000
+                application_progress_error['dimensions'] = self.dimensions
+
+                print application_progress_error['value']
+
+                self.monasca.send_metrics([application_progress_error])
+
+            time.sleep(MONITORING_INTERVAL)
+
+
+    def _get_elapsed_time(self, gmt_timestamp):
+        try:
+            local_tz = tzlocal.get_localzone()
+
+        except Exception as e:
+            local_tz = "America/Recife"
+            local_tz = pytz.timezone(local_tz)
+
+        date_time = datetime.strptime(gmt_timestamp, '%Y-%m-%dT%H:%M:%S.%fGMT')
+        date_time = date_time.replace(tzinfo=pytz.utc).astimezone(local_tz)
+        date_time = date_time.replace(tzinfo=None)
+        datetime_now = datetime.now()
+        elapsed_time = datetime_now - date_time
+
+        return elapsed_time.seconds
+
+    def _discover_id_from_spark(self):
+        for i in range(30):
```

Comment:
Is 30 the number of attempts? If yes, create a generic parameter to this.

Reply:
I agree.